### PR TITLE
Streaming flow control

### PR DESCRIPTION
**Pull request by ashione**
## Why are these changes needed?
Support flow control for channel transfer.
A flow control method is implemented in this PR and its main functions is:
* Avoid overstocking or increasing checkpoint completion time caused by too much data when downstream consumption is slow.
*  Speed checkpoint completion up via passing by barrier

How to do flow control:
WriterConsumedStep and ReaderConsumedStep are proposed by unconsumed sequence flow control. 
When the upstream writes data to the downstream, it maintains a sequence of unconsumped data. The maximum length of this sequence is `WriterConsumedStep`.
And each time the downstream consumption data reaches a certain step (named `ReaderConsumedStep`), it will notify the upstream of the offset at this time. Or the current consumption is the last data. In order to avoid the omission of consumption offsets, offset information will also be sent to the upstream.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by ashione** [comment]
@MissiontoMars @raulchen @zhijunfu @jovany-wang Please help use to review this PR in your spare time.

File: streaming/src/channel.h [hunk 1]
```diff
@@ -56,6 +56,8 @@ struct ConsumerChannelInfo {
   uint64_t last_queue_target_diff;
   uint64_t get_queue_item_times;
   ActorID actor_id;
+  // Total count of notify request.
+  uint64_t notify_cnt;
```
Comment (zhijunfu) [inline streaming/src/channel.h]: not specific to this PR, but it's good to do default initialization for native types.

File: streaming/src/config/streaming_config.h [hunk 2]
```diff
@@ -46,6 +55,12 @@ class StreamingConfig {
   DECL_GET_SET_PROPERTY(uint32_t, EmptyMessageTimeInterval, empty_message_time_interval_)
   DECL_GET_SET_PROPERTY(streaming::proto::OperatorType, OperatorType, operator_type_)
   DECL_GET_SET_PROPERTY(const std::string &, JobName, job_name_)
+  DECL_GET_SET_PROPERTY(uint32_t, WriterConsumedStep, writer_consumed_step_)
+  DECL_GET_SET_PROPERTY(uint32_t, ReaderConsumedStep, reader_consumed_step_)
+  DECL_GET_SET_PROPERTY(streaming::proto::FlowControlType, FlowControlType,
+                        flow_control_type_)
+  DECL_GET_SET_PROPERTY(uint32_t, EventDrivenFlowControlInterval,
+                        event_driven_flow_control_interval)
```
Comment (zhijunfu) [inline streaming/src/config/streaming_config.h]: should this be event_driven_flow_control_interval_?

File: streaming/src/config/streaming_config.h [hunk 3]
```diff
@@ -33,6 +33,15 @@ class StreamingConfig {
 
   std::string task_job_id_ = JobID::Nil().Hex();
 
+  streaming::proto::FlowControlType flow_control_type_ =
+      streaming::proto::FlowControlType::UnconsumedSeq;
```
Comment (zhijunfu) [inline streaming/src/config/streaming_config.h]: would be nice to add some description on flow control and the types.

File: streaming/src/flow_control.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  virtual bool ShouldFlowControl(ProducerChannelInfo &channel_info) = 0;
+};
+
+class NoFlowControl : public FlowControl {
+ public:
+  bool ShouldFlowControl(ProducerChannelInfo &channel_info) { return false; }
+  ~NoFlowControl() = default;
+};
+
+class UnconsumedSeq : public FlowControl {
```
Comment (zhijunfu) [inline streaming/src/flow_control.h]: maybe `UnconsumedSeqFlowControl`?

File: streaming/src/protobuf/streaming.proto [hunk 5]
```diff
@@ -11,6 +11,12 @@ enum OperatorType {
   SINK = 3;
 }
 
+enum FlowControlType {
+  UNKNOWN_FLOW_CONTROL_TYPE = 0;
```
Comment (zhijunfu) [inline streaming/src/protobuf/streaming.proto]: where is this used?

File: streaming/src/config/streaming_config.cc [hunk 6]
```diff
@@ -37,6 +37,21 @@ void StreamingConfig::FromProto(const uint8_t *data, uint32_t size) {
   if (config.empty_message_interval() != 0) {
     SetEmptyMessageTimeInterval(config.empty_message_interval());
   }
+  if (config.flow_control_type() != proto::FlowControlType::UNKNOWN_FLOW_CONTROL_TYPE) {
+    SetFlowControlType(config.flow_control_type());
+  }
+  if (config.writer_consumed_step()) {
```
Comment (MissiontoMars) [inline streaming/src/config/streaming_config.cc]: `if (config.writer_consumed_step() != 0)` , keep the same with `config.empty_message_interval `

File: streaming/src/data_reader.cc [hunk 7]
```diff
@@ -199,10 +199,9 @@ StreamingStatus DataReader::GetMergedMessageBundle(std::shared_ptr<DataBundle> &
 
 StreamingStatus DataReader::GetBundle(const uint32_t timeout_ms,
                                       std::shared_ptr<DataBundle> &message) {
-  // Notify consumed every item in this mode.
+  // Notify last fetched item consumed.
```
Comment (MissiontoMars) [inline streaming/src/data_reader.cc]: // Notify last fetched item ?

File: streaming/src/data_writer.cc [hunk 8]
```diff
@@ -447,5 +431,54 @@ void DataWriter::EmptyMessageTimerCallback() {
   }
 }
 
+void DataWriter::NotifyConsumedByRefresh(ProducerChannelInfo &channel_info) {
```
Comment (MissiontoMars) [inline streaming/src/data_writer.cc]: Is `NotifyConsumedByRefresh` means `refresh` and then  `NotifyConsumedItem`?
Maybe better to use `RefreshAndNotifyConsumed`?

File: streaming/src/data_reader.h [hunk 9]
```diff
@@ -103,6 +103,9 @@ class DataReader {
   ///
   void NotifyConsumedItem(ConsumerChannelInfo &channel_info, uint64_t offset);
 
+  //// Notify input queues to clear data who channel of message belongs to.
```
Comment (MissiontoMars) [inline streaming/src/data_reader.h]: Typo. who -> whose

File: streaming/src/data_writer.cc [hunk 10]
```diff
@@ -447,5 +431,54 @@ void DataWriter::EmptyMessageTimerCallback() {
...
+                    channel_info.writer_ring_buffer->IsFull()};
+        event_service_->Push(event);
+        ++channel_info.flow_control_cnt;
+      }
+    }
+    std::this_thread::sleep_for(MockTimer);
+  }
+}
+
+void DataWriter::GetOffsetInfo(
```
Comment (MissiontoMars) [inline streaming/src/data_writer.cc]: Is this function just for test?

File: streaming/src/data_writer.h [hunk 11]
```diff
@@ -30,6 +31,40 @@ namespace streaming {
...
+                       const std::vector<uint64_t> &queue_size_vec);
+
+  ///  To increase throughout, we employed an output buffer for message transformation,
+  ///  which means we merge a lot of message to a message bundle and no message will be
+  ///  pushed into queue directly util daemon thread does this action.
+  ///  Additionally, writing will block when buffer ring is full intentionly.
+  ///  \param q_id
+  ///  \param data
+  ///  \param data_size
+  ///  \param message_type
```
Comment (MissiontoMars) [inline streaming/src/data_writer.h]: Add more comment for these parameters.

File: streaming/src/data_writer.h [hunk 12]
```diff
@@ -79,42 +104,25 @@ class DataWriter {
...
 
  private:
   std::shared_ptr<EventService> event_service_;
 
   std::shared_ptr<std::thread> empty_message_thread_;
+
+  std::shared_ptr<std::thread> flow_control_thread_;
   // One channel have unique identity.
   std::vector<ObjectID> output_queue_ids_;
+  // Flow controller makes decision when it's should be blocked and avoid
```
Comment (MissiontoMars) [inline streaming/src/data_writer.h]: The flow controller makes a decision

File: streaming/src/flow_control.cc [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+        channel_info.queue_info.consumed_seq_id + consumed_step_;
+    STREAMING_LOG(DEBUG) << "Flow control stop writing to downstream, current max id => "
+                         << channel_info.current_seq_id << ", target seq id => "
+                         << queue_info.target_seq_id << ", consumed_id => "
+                         << queue_info.consumed_seq_id << ", q id => "
+                         << channel_info.channel_id
+                         << ". if this log keeps printing, it means something wrong "
+                            "with queue's info API, or downstream node is not "
+                            "consuming data.";
+    // Double check after refreshing.
```
Comment (MissiontoMars) [inline streaming/src/flow_control.cc]: Add more comment about why we need double checking ?

File: streaming/src/flow_control.h [hunk 14]
```diff
@@ -0,0 +1,45 @@
+#ifndef RAY_STREAMING_FLOW_CONTROL_H
+#define RAY_STREAMING_FLOW_CONTROL_H
+#include "channel.h"
+
+namespace ray {
+namespace streaming {
+class ProducerTransfer;
+/// We devise a flow control system in queue channel, and that's called flow
+/// control by unconsumed seq. Upstream worker will detect consumer stat via
```
Comment (MissiontoMars) [inline streaming/src/flow_control.h]: Typo stat -> stats/status

File: streaming/src/flow_control.h [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+/// api so it can keep fixed length messages in this process, which make a
```
Comment (MissiontoMars) [inline streaming/src/flow_control.h]: makes

File: streaming/src/protobuf/streaming.proto (hunk 5 above)
Comment (ashione) [reply streaming/src/protobuf/streaming.proto]: The first of enum list is default value in protobuf generating.

File: streaming/src/data_writer.cc (hunk 10 above)
Comment (ashione) [reply streaming/src/data_writer.cc]: > Is this function just for test?

For checkpoint, but it's can be used in testing.

File: streaming/src/config/streaming_config.cc [hunk 16]
```diff
@@ -37,6 +37,21 @@ void StreamingConfig::FromProto(const uint8_t *data, uint32_t size) {
   if (config.empty_message_interval() != 0) {
     SetEmptyMessageTimeInterval(config.empty_message_interval());
   }
+  if (config.flow_control_type() != proto::FlowControlType::UNKNOWN_FLOW_CONTROL_TYPE) {
```
Comment (chaokunyang) [inline streaming/src/config/streaming_config.cc]: Please add flow_control_type in java/python too

File: streaming/src/flow_control.h (hunk 4 above)
Comment (ashione) [reply streaming/src/flow_control.h]: Good suggestion.
Fixed.

> maybe `UnconsumedSeqFlowControl`?

File: streaming/src/config/streaming_config.cc (hunk 16 above)
Comment (ashione) [reply streaming/src/config/streaming_config.cc]: Flow control config have been parsed in python/java worker.

**Review by MissiontoMars — APPROVED** [review]

**Review by chaokunyang — APPROVED** [review]

**Comment by jovany-wang** [comment]
CI seems unrelated. I'm merging this.

**Pull request merged**